components/scream/src/physics/p3/p3_cloud_sed_impl.hpp

```diff
@@ -98,17 +98,17 @@ ::cloud_sedimentation(
 	    //between nc_incld and nc
             nc(pk).set(qc_gt_small, nc_incld(pk)*cld_frac_l(pk));
 
-	          dum = 1 / pow(lamc(pk), bcn);
+            dum = 1 / pow(lamc(pk), bcn);
             V_qc(pk).set(qc_gt_small, acn(pk)*tgamma(4 + bcn + mu_c(pk)) * dum / tgamma(mu_c(pk)+4));
             if (do_predict_nc) {
               V_nc(pk).set(qc_gt_small, acn(pk)*tgamma(1 + bcn + mu_c(pk)) * dum / tgamma(mu_c(pk)+1));
             }
-
-            const auto Co_max_local = max(qc_gt_small, -1,
-                                          V_qc(pk) * dt_left * inv_dz(pk));
-            if (Co_max_local > lmax)
-              lmax = Co_max_local;
           }
```

Comment:
This was an actual bug in the c++ implementation. I discovered by accident when using better random values in the bfb unit test of cloud_sed.

Reply:
yikes!

Reply:
Was the bug that Co_max was < 0 if `!qc_gt_small.any()`? If so, that's b/c the team reduction returns `-DBL_MAX` (or `-FLT_MAX` for single precision) if the reduction variable isn't updated. In the micro-app, we handled this situation differently, demonstrated [here](https://github.com/E3SM-Project/scream-docs/blob/e83980e34c51b19a24d2209f7bd7785f3d483dfb/micro-apps/p3/p3_final_impl.hpp#L212). The idea is to save computation at every pk by instead just doing one test at the end for `Co_max < 0`.

Reply:
If that is indeed the bug, you could instead return the local test to the inside of the any() block and add a fast exit to generalized_sedimentation on Co_max < 0.

Reply:
@ambrad, the bug was that, when qc_gt_small.none(), it needed to skip the Co_max stuff, because that's what fortran does.

Reply:
@ambrad , whoops, the exact opposite of what I just said. We need to always do the Co_max stuff, even when qc_gt_small.none()

Reply:
@jgfouca I agree. But that doesn't necessarily imply Co_max should be updated always. For future reference, I'm going to explain here as long as I looked at this again, but you needn't do anything in this PR.

I believe the bug you caught is that Co_max = -DBL_MAX when it is passed to generalized_sedimentation. That value is because Kokkos's team reduction ignores Co_max's initial value of 0 and rather initializes the reduction variable with -DBL_MAX. If !qc_get_small.any(), Co_max emerges as -DBL_MAX, when we want it to be 0.

Now, if Co_max is 0, then all the V_* velocities are 0. Then fluxdiv is everywhere 0 in the transport routine, so there is no change to the advected tracer, not even in a BFB sense, because the routine just adds and subtracts 0s.

In the micro-app -- line linked in my previous comment -- we handled the situation more efficiently. If Co_max < 0 on exit of the loop, we recognized this is because the loop never did anything; it has the value initialized by Kokkos. In the micro-app, we then broke from the overall advection loop because once a tracer is everywhere below the 'small' threshold, it will remain everywhere fixed in space thereafter.

There is a related issue that persists, though the main bug was fixed a couple of months ago. There are _incld variants of the tracers now; for a tracer q, q <= q_incld. ktop and kbot are found based on q >= qsmall, but then the V code chunk runs based on q_incld >= qsmall, which can be true when q < qsmall. So the range [ktop,kbot] is potentially a subset of the range implied by q_incld >= qsmall.

Reply:
Made an issue for this.